drivers/dht/dht.c

```diff
@@ -19,192 +20,267 @@
+#define SPIN_TIMEOUT            (85U)
+/* According to datasheet the DHT sensor cannot measure more than once
+ * every 2 seconds but most are faster */
+#define RESPAWN_TIMEOUT         (2U * US_PER_SEC)
+/* The start signal by pulling data low for at least 18 ms for DHT11, at
+ * most 20 ms (AM2301 / DHT22 / DHT21). Then release the bus and the
+ * sensor should respond by pulling data low for 80 µs, then release for
+ * 80µs before start sending data. */
+#define START_LOW_TIME          (19U * US_PER_MS)
+#define START_THRESHOLD         (75U)
+/* DHTs have to wait for power 1 or 2 seconds depending on the model */
+#define POWER_WAIT_TIMEOUT      (2U * US_PER_SEC)
+
+enum {
+    BYTEPOS_HUMIDITY_HIGH = 0,
+    BYTEPOS_HUMIDITY_LOW = 1,
+    BYTEPOS_TEMPERATURE_HIGH = 2,
+    BYTEPOS_TEMPERATURE_LOW = 3,
+    BYTEPOS_CHECKSUM = 4,
+};
+
+struct dht_data {
+    gpio_t pin;
+    gpio_mode_t in_mode;
+    uint8_t data[5];
+    int8_t bit_pos;
+    uint8_t bit;
+};
+
+static inline void _wait_for_level(gpio_t pin, bool expected, uint32_t start)
 {
+    uint8_t pre_timeout = 0;
+    while ((gpio_read(pin) != expected) && (++pre_timeout
+        || ztimer_now(ZTIMER_USEC) < start + SPIN_TIMEOUT)) {}
+}
+
+static inline int8_t _send_start_signal(dht_t *dev)
+{
+    uint32_t start;
     gpio_init(dev->params.pin, GPIO_OUT);
+    gpio_clear(dev->params.pin);
+    ztimer_sleep(ZTIMER_USEC, START_LOW_TIME);
+    /* sync on device */
     gpio_set(dev->params.pin);
+    gpio_init(dev->params.pin, dev->params.in_mode);
+    /* check device response (80 µs low then 80 µs high) */
+    start = ztimer_now(ZTIMER_USEC);
+    _wait_for_level(dev->params.pin, 0, start);
+    if (ztimer_now(ZTIMER_USEC) - start > START_THRESHOLD) {
+        DEBUG_PUTS("[dht] error: response low pulse > START_THRESHOLD");
+        return -ENODEV;
+    }
+    _wait_for_level(dev->params.pin, 1, start);
+    start = ztimer_now(ZTIMER_USEC);
+    _wait_for_level(dev->params.pin, 0, start);
+    if (ztimer_now(ZTIMER_USEC) - start < START_THRESHOLD) {
+        DEBUG_PUTS("[dht] error: response high pulse < START_THRESHOLD");
+        return -ENODEV;
+    }
+    return 0;
 }
 
-/**
- * @brief   Wait until the pin @p pin has level @p expect
- *
- * @param   pin     GPIO pin to wait for
- * @param   expect  Wait until @p pin has this logic level
- * @param   timeout Timeout in µs
- *
- * @retval  0       Success
- * @retval  -1      Timeout occurred before level was reached
- */
-static inline int _wait_for_level(gpio_t pin, bool expect, unsigned timeout)
+static inline void _bit_parse(struct dht_data *arg)
 {
-    while (((gpio_read(pin) > 0) != expect) && timeout) {
-        xtimer_usleep(1);
-        timeout--;
+    int8_t pos = arg->bit_pos++;
+    if (arg->bit) {
+        arg->data[pos / 8] |= (0x80U >> (pos % 8));
     }
+}
 
-    return (timeout > 0) ? 0 : -1;
+static inline void _busy_wait_read(struct dht_data *arg)
```

Comment:
```suggestion
static void _busy_wait_read(struct dht_data *arg)
```

Modern compilers ignore the `inline` anyway and do inline `static` functions without `inline` (where deemed beneficial) but also may not inline `static inline` functions when that is better aligned with the optimization goal. Given that `static` functions do not need to be externally callable, the compiler can do some aggressive optimizations on the function call overhead - so a not inlined `static` "function call" can be much less costly than a real function call of an external function.

But this is a style question, so do as you prefer. If you agree, you could also change the other functions.